src/security/noise/noise_connection.cpp

```diff
@@ -99,9 +100,14 @@ namespace libp2p::connection {
 
   void NoiseConnection::write(gsl::span<const uint8_t> in, size_t bytes,
                               libp2p::basic::Writer::WriteCallbackFunc cb) {
+    if (0 == plaintext_len_to_write_) {
+      plaintext_len_to_write_ = bytes;
+    }
     if (bytes == 0) {
-      auto n{already_wrote_};
+      BOOST_ASSERT(already_wrote_ >= plaintext_len_to_write_);
+      auto n{plaintext_len_to_write_};
       already_wrote_ = 0;
+      plaintext_len_to_write_ = 0;
```

Comment:
```c++
```plaintext_len_to_write_``` can not be a member because of frame sends. ```write``` can be called again before all the frames will be sent, and ```plaintext_len_to_write_``` become invalid. It seems better to push ```plaintext_len_to_write_``` as a lambda context member.